Create new model for python packages

asmacdo opened the pull request:
Please note, this is being merged to a feature branch which will contain this commit and others to get sync and publish working. Because this is only changing the model, it is expected that any tests outside of `plugins/test/unit/plugins/test_models.py` may fail.

The filename contains the name, version, platform, and python versions.
This means that filename is guaranteed to be unique on PyPI.
Additionally, this commit removes some unecessary field names to prevent
them from being wrongly populated with data from a later release, which
will occur if the packages are built from the JSON metadata from PyPI.

closes #1882

jeremycline commented on plugins/pulp_python/plugins/migrations/0002_smaller_model_for_json.py:
```diff
@@ -0,0 +1,20 @@
+"""
+This migration removes `home_page`, `platform`, and `license` from the units_python_package
+collection. Additionally, it renames `_filename` to `filename` and makes it the pk.
```
It eventually clicked that pk was primary key, but I'd prefer it to be spelled out for me.

jeremycline commented on plugins/pulp_python/plugins/models.py:
```diff
@@ -1,120 +1,85 @@
     summary = StringField()
 
     _checksum = StringField()
     _checksum_type = StringField(default=DEFAULT_CHECKSUM_TYPE)
```
I guess it's outside the scope of this PR, but I really wish this was just "checksum" and in the format "<type>:<hex_value>". It's done this way on all the other units, though, so I suppose we should be consistent.

jeremycline commented on plugins/pulp_python/plugins/models.py:
```diff
@@ -1,120 +1,85 @@
 
 class Package(FileContentUnit):
     """
     This class represents a Python package.
```
It would be great if you documented this class more extensively. I know why you are using the filename as the unit key, but only because you told me in person. I'd like this doc-block to explain what each field is, where it comes from, and why it is there. You've done a lot of research I don't want to have to do myself :smile:

asmacdo commented on plugins/pulp_python/plugins/models.py:
```diff
@@ -1,120 +1,85 @@
 
 from mongoengine import StringField
 from pulp.server.db.model import FileContentUnit
+import twine  # noqa
```
@jeremycline am I crazy? I had to add noqa here because flake8 said this was the incorrect order...

jeremycline replied:
That is weird. I feel like someone else had a similar issue with import ordering recently, but I can't recall who.

jeremycline commented on plugins/pulp_python/plugins/models.py:
```diff
@@ -1,120 +1,85 @@
+        Create and return (but do not save) an instance of a single python package object from the
+        metadata available from PyPI JSON.
+
+        :param file_data: metadata specific to a file
```
I'm not exactly sure what this means. Which file? At first I thought it was from the JSON from the PyPI API, but I see there's a second dictionary, distribution_data. How do they differ and where do they come from?

jeremycline commented on plugins/pulp_python/plugins/models.py:
```diff
@@ -1,120 +1,85 @@
+        """
+        package_attrs = {}
+        package_attrs['filename'] = file_data['filename']
+        package_attrs['name'] = distribution_data['name']
```
Couldn't you just drop ``name`` into the list you're iterating over below?

jeremycline commented on plugins/pulp_python/plugins/models.py:
```diff
@@ -1,120 +1,85 @@
+        """
+
+        meta_dict = twine.package.PackageFile.from_filename(path, comment='').metadata_dictionary()
+        filtered_dict = {}
```
I dream of the day we can use dictionary comprehensions.

jeremycline commented:
Just documentation comments, really. If you want me to proof-read your docs I am happy to do that, otherwise merge when you're ready.

asmacdo commented:
@jeremycline, I added more docs, removed a couple more fields. I would appreciate if you would let me know if the new documentation makes sense and sounds rational. I kept the commit separate to make it easy for you to see. I'll rebase before merge.

jeremycline commented on plugins/pulp_python/plugins/models.py:
```diff
@@ -8,20 +8,54 @@
     """
     This class represents a Python package.
+
+    Packages stored on PyPI have significantly more metadata associated with them but we have
```
Comma before the "but"

jeremycline commented on plugins/pulp_python/plugins/models.py:
```diff
@@ -8,20 +8,54 @@
+
+    Package - An individual, installable python package. A package is of a specific version, and
+              may also require a specific architecture and python version. Additionally, there
+              may be seperate packages for each package type (sdist, whl, zip).
```
Maybe:
```
Package - An individual, installable Python package. A package is defined by the version of
    the project, target architecture, and Python version. Note that there may not be a target
    architecture in the case of architecture-agnostic packages. A Python version is also
    optional if there are no requirements for a specific runtime version. There are three
    packaging formats: sdist, whl, and zip.
```
I'm not convinced the above is better, so pick and choose as you like. If you stick with what you have, ``s/seperate/separate``

jeremycline commented on plugins/pulp_python/plugins/models.py:
```diff
@@ -8,20 +8,54 @@
+    :ivar name: name of the project, ex. scipy
+    :type name: basestring
+    :ivar version: Contains the distribution's version number. This field must be in the format
+                   specified in PEP 386 .
```
There's a space before the period.

jeremycline commented on plugins/pulp_python/plugins/models.py:
```diff
@@ -8,20 +8,54 @@
+    :ivar version: Contains the distribution's version number. This field must be in the format
+                   specified in PEP 386 .
+    :type version: basestring
+    :ivar filename: name of the file containaing the package and metadata
```
s/containaing/containing
